File: features/5_games/sweeper.js
let math = require('math');
let games = {};
const def_size = 5;
const mineamount = 0.2;
let mta = {}; //minetypaamount
const outsize = 5;

function isNumeric(n){return !isNaN(parseFloat(n)) && isFinite(n);}

function printboard(board,amount){
  let out = "";
  for(let x in board){
    let outt = "";
    for(let y in board[x]){
      if(board[x][y]==-1){
        let sum = [0,0];
        for(let x1 = -1;x1<2;x1++)
          for(let y1=-1;y1<2;y1++)
            if(x-x1>-1&&y-y1>-1&&x-x1<size&&y-y1<size)
            if(board[x-x1][y-y1]>0){
              sum[0] += math.cos(board[x-x1][y-y1]*6.283185/amount);
              sum[1] += math.sin(board[x-x1][y-y1]*6.283185/amount);
            }
        let real = (math.round(sum[0]*1000)/1000).toString();
        let imag = (math.round(sum[1]*1000)/1000).toString();
        if(real=='0')
          out += '`'+" ".repeat(outsize)+'` ';
        else
          out +="`"+real.slice(0,outsize)+" ".repeat(outsize-real.length)+"` ";
        if(imag=='0')
          outt += '`'+" ".repeat(math.floor(outsize))+'` ';
        else 
          outt+="`"+imag.slice(0,outsize)+" ".repeat(math.floor(outsize-imag.length))+"` ";
      }else{
        out += '`'+"#".repeat(outsize)+'` '
        outt += '`'+"#".repeat(outsize)+'` '
      }
    }
    out += "\n"+outt+"\n\n";
  }
  return out;
}

exports.run = (msg, client) => {
  if(!msg.content.toLowerCase().startsWith('m!')) return;
  let args = msg.content.split(' ').slice(1);
  
  let player = {"board":games[msg.author.id],"mta":mta[msg.author.id]};
  let player = JSON.parse(client.db.v2getuser("sweeper",msg.author.id))
  switch(msg.content.split(' ')[0].slice(2).toLowerCase()){
    case 'start':{
      if(args.length<1){msg.reply('too little arguments');break}
      size = args[1]??def_size
      let board = [];
      for(let x=0;x<size;x++){
        board.push([])
        for(let y=0;y<size;y++)
          if(math.random()<mineamount)
            board[x].push(math.floor(math.random()*args[0])+1)
          else
            board[x].push(0)
      }
      player = {};
      player.board = board;
      player.mta = args[0];
      msg.reply('started probably idk')
    break}
    case 'sweep':{
      if(!player){msg.reply('you have to start a game!');break}
      if(args.length<2){msg.reply('too few agruments');break;}
      if(!isNumeric(args[0])||!isNumeric(args[1])){msg.reply('expected 2 intigers');break}
      let [x,y] = args;
      if(player.board[x][y]>0){
        msg.reply('you clicked a mine!')
        delete player.board;
        delete player.mta;
      }else{
        player.board[x][y] = -1;
        msg.reply(printboard(player.board,player.mta))
      }
    break}
    case 'help':{
      msg.reply('welcome to the complex minesweeper!\nin this minesweeper, there are N different mine types\neach mine\'s value is one of the Nths roots of unity\nthe values on the cells are the sum of values of the neighbouring 8 cells(empty=0)\ntop is the real part, bottom is the imaginary part\n\ncommands:\n[]-> optional,\n> m!start mine_type_amount [board_size=5]\n\n> m!sweep x y\nnote: 0-indexed, so for a 5x5 board the possibilities are 0,1,2,3,4')
    break}
  };
  if(player=={}) client.db.v2deluser("sweeper",msg.author.id);
  else client.db.v2setuser("sweeper",msg.author.id,JSON.stringify(player))
}

File: features/5_games/sweeper.py
"""This is a minesweeper working differently than other minesweeper\n
The catch is that what's shown is using cosinu, sinus and rounding\n
Now please give me size !!!"""
import math

DEF_SIZE = 5
MINEAMOUNT = 0.2
OUTSIZE = 5

isNumeric = math.isfinite
"""
not used as this exists due to js implementation
"""

def printboard(board, amount):
    "Prints the board"
    out = ""
    for x in range(len(board)):
        outt=""
        for y in range(len(board[x])):
            if board[x][y] == -1:
                sum = [0, 0]
                for x1 in range(-1, 2):
                    for y1 in range(-1, 2):
                        if x-x1>-1 and y-y>-1 and x-x1<size and y-y1<size:
                            if board[x - x1][y - 1]>0:
                                sum[0] += math.cos(board[x-x1][y-y1]*6.283185/amount)
                                sum[1] += math.sin(board[x-x1][y-y1]*6.283185/amount)
                real = str(round(sum[0], 4))
                imag = str(round(sum[1], 4))
                if real=='0':
                    out += '`'+" "*OUTSIZE+'` '
                else:
                    out += '`'+real[0:OUTSIZE]+" "*(OUTSIZE-len(real))+'` '
                if imag=='0':
                    outt += '`'+" "*math.floor(OUTSIZE)+'` '
                else:
                    outt += '`'+real[0:OUTSIZE]+" "*(math.floor(OUTSIZE-len(real)))+'` '
            else:
                out += '`'+"#"*OUTSIZE+'` '
                outt += '`'+"#"*OUTSIZE+'` '
        out += "\n"+outt+"\n\n"
    return out

import json
def run(msg, client):
    "Run the minesweeper"
    if not msg.content.lower().startswith('m!'):
        return
    args = msg.content.split(' ')[1:]
    player = json.loads(client.db.v2getuser("sweeper",msg.author.id))
    switch = {}
    def case():
        if len(args) < 1:
            msg.reply('too little arguments')
            return
        size = args[1] or DEF_SIZE
        board = []
        import random
        for x in range(size):
            board.append([])
            for _ in range(size):
                if random.random()<MINEAMOUNT:
                    board[x].append(math.floor(random.random()*args[0])+1)
                else:
                    board[x].append(0)
        player = {}
        player["board"] = board
        player["mta"] = args[0]
        msg.reply('started probably idk') #I'm scared if even 0lie doesn't know if the start function works
        return
    switch["start"] = case
    def case():
        if not player:
            msg.reply('you have to start a game!')
            return
        if len(args) < 2:
            msg.reply('too few agruments')
            return
        if (not math.isfinite(args[0]) or not math.isfinite(args[1])): #Might need to convert first
            msg.reply('expected 2 intigers') #Integers not intigers
            return
        x, y = args[:2]
        if player["board"][x][y]>0:
            msg.reply('you clicker a mine!')
            del player["board"]
            del player["mta"]
        else:
            player["board"][x][y] = -1
            msg.reply(printboard(player["board"],player["mta"]))
        return
    switch["sweep"] = case
    def case():
        msg.reply('welcome to the complex minesweeper!\nin this minesweeper, there are N different mine types\neach mine\'s value is one of the Nths roots of unity\nthe values on the cells are the sum of values of the neighbouring 8 cells(empty=0)\ntop is the real part, bottom is the imaginary part\n\ncommands:\n[]-> optional,\n> m!start mine_type_amount [board_size=5]\n\n> m!sweep x y\nnote: 0-indexed, so for a 5x5 board the possibilities are 0,1,2,3,4')
        return
    switch["help"] = case
    del case
    switch[msg.content.split(' ')[0][2:].lower()]()
    del switch
    if player == {}:
        client.db.v2deluser("sweeper",msg.author.id)
    else:
        client.db.v2setuser("sweeper",msg.author.id,json.dumps(player))
